godel_robots/abb/abb_file_suite/rapid_generator/src/rapid_emitter.cpp

```diff
@@ -0,0 +1,183 @@
+  // for lengthFreeMotion to end of points, emit grind moves
+  for (std::size_t i = startProcessMotion; i < endProcessMotion; ++i)
+  {
+    if (i == startProcessMotion)
+    {
+      emitGrindMotion(os, params, i, true, false);
+    }
+    else if (i == endProcessMotion-1)
+    {
+      emitGrindMotion(os, params, i, false, true);
+    }
+    else
+    {
+      emitGrindMotion(os, params, i);
+    }
+  }
+  
+  emitSetOutput(os, params, 0);
+  
+  // for lengthFreeMotion to end of points, emit grind moves
+  for (std::size_t i = endProcessMotion; i < points.size(); ++i)
+  {
+    if (i == endProcessMotion) emitFreeMotion(os, params, i, 0.0, true);
+    else if (i == (points.size() - 1)) emitFreeMotion(os, params, i, points[i].duration_, true);
+    else emitFreeMotion(os, params, i, points[i].duration_, false);
+  }
+
+  os << "EndProc\n";
+  // write any footers including main procedure calling the above
+  os << "ENDMODULE\n";
+  return true;
+}
+
+
+bool rapid_emitter::emitGrindMotion(std::ostream& os, const ProcessParams& params, size_t n, bool start, bool end)
+{
+  if (params.wolf)
+  {
+    if (start)
+    {
+      os << "GrindLStart CalcRobT(jTarget_" << n << ",tool0), v100, gr1, fine, tool0;\n";
+    }
+    else if (end)
+    {
+      os << "GrindLEnd CalcRobT(jTarget_" << n << ",tool0), v100, fine, tool0;\n";
+    }
+    else
+    {
+      os << "GrindL CalcRobT(jTarget_" << n << ",tool0), v100, z40, tool0;\n";
+    }
+  }
+  else
+  {
+    os << "MoveL CalcRobT(jTarget_" << n << ",tool0), vProcessSpeed, z5, tool0;\n";
+  }
+  return os.good();
+}
+
+bool rapid_emitter::emitFreeMotion(std::ostream& os, const ProcessParams& params, size_t n, double duration, bool stop_at)
+{
+  const char* zone = stop_at ? "fine" : "z20";
+
+  if (duration <= 0.0)
+  {
+    os << "MoveJ CalcRobT(jTarget_" << n << ",tool0), vMotionSpeed," << zone << ", tool0;\n";
+  }
+  else
+  {  
+    os << "MoveJ CalcRobT(jTarget_" << n << ",tool0), vMotionSpeed, \\T:=" << duration << ", " << zone << ", tool0;\n";
+  }
+  return os.good();
+}
+bool rapid_emitter::emitJointPosition(std::ostream& os, const TrajectoryPt& pt, size_t n)
+{
+  os << "TASK PERS jointtarget jTarget_" << n << ":=[[";
+    for (size_t i = 0; i < pt.positions_.size() ; i++)
+    {
+      os << pt.positions_[i];
+      if (i < pt.positions_.size()-1)
+      {
+        os << ",";
+      }
+    }
+    
+  os << "],[9E9,9E9,9E9,9E9,9E9,9E9]];\n";
+  return true;
+}
+
+bool rapid_emitter::emitSetOutput(std::ostream& os, const ProcessParams& params, size_t value)
+{
+  if (params.wolf == false)
+  {
+    os << "WaitTime\\InPos, 0.01;\n";
+    os << "SETDO " << params.output_name << ", " << value << ";\n";
+    // os << "SETDO " << "EX_Vlv_1A_12" << ", " << value << ";\n";
+    // os << "SETDO " << "EX_Vlv_1B_13" << ", " << value << ";\n";
+  }
+  return os.good();
+} 
+
+bool rapid_emitter::emitProcessDeclarations(std::ostream& os, const ProcessParams& params, size_t value)
+{
+  if (params.wolf)
```

Comment:
I need to explain this. It interacts with Wolf Robotics proprietary software if available.